src/plugins/external_tool_support/src/ExternalToolManager.cpp

```diff
@@ -193,28 +190,42 @@ QString ExternalToolManagerImpl::addTool(ExternalTool *tool) {
         foreach (const QString &dependency, toolDependencies) {
             dependencies.insertMulti(dependency, tool->getId());
         }
+    }
+}
+
+void ExternalToolManagerImpl::addToolToListsIfNecessary(const ExternalTool &tool) {
+    if (!tool.isValid()) {
+        if (tool.getDependencies().isEmpty() || dependenciesAreOk(tool.getId())) {
+            validateList << tool.getId();
+        }
+    }
+
+    if (!validateList.contains(tool.getId()) && !tool.isModule() && !tool.isValid()) {
+        searchList << tool.getId();
+    }
+}
 
-        if (dependenciesAreOk(tool->getId()) && !tool->isValid()) {
-            if (tool->isModule()) {
-                QString masterId = tool->getDependencies().first();
+QString ExternalToolManagerImpl::findToolPath(const ExternalTool &tool) {
+    QString toolPath;
+    SAFE_POINT(nullptr != etRegistry, "The external tool registry is nullptr", toolPath);
+
+    QStringList toolDependencies = tool.getDependencies();
+    if (!toolDependencies.isEmpty()) {
+        if (dependenciesAreOk(tool.getId()) && !tool.isValid()) {
+            if (tool.isModule()) {
+                QString masterId = tool.getDependencies().first();
                 ExternalTool *masterTool = etRegistry->getById(masterId);
-                SAFE_POINT(masterTool, QString("An external tool '%1' isn't found in the registry").arg(masterId), "");
+                SAFE_POINT(nullptr != masterTool, QString("An external tool '%1' isn't found in the registry").arg(masterId), toolPath);
 
                 toolPath = masterTool->getPath();
             } else {
-                toolPath = tool->getPath();
+                toolPath = tool.getPath();
             }
-            validateList << tool->getId();
+        } else {
```

Comment:
Empty else?

Reply:
I don't understand, how can I show that all states are covered and have no missed states (https://github.com/ugeneunipro/ugene/pull/152#pullrequestreview-535011600). Can I remove empty else branches? And you and other developers will believe that the block is missing for a reason.

Reply:
Once you will fix the method contract or name for 'findToolPath' it will be much easier to understand why in some cases you return the tool path and in some reason an empty string  (empty if).